oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/PersistenceTimer.java

```diff
@@ -92,69 +105,111 @@ public void start(ModuleManager moduleManager, CoreModuleConfig moduleConfig) {
+                            continue;
+                        }
 
-                if (debug) {
-                    log.info("build batch persistence duration: {} ms", System.currentTimeMillis() - startTime);
-                }
-            } finally {
-                timer.finish();
-            }
+                        List<PrepareRequest> prepareRequestList = this.prepareRequests.subList(
+                            0, Math.min(maxSyncoperationNum, this.prepareRequests.size()));
+                        partition = new ArrayList<>(prepareRequestList);
+                        prepareRequestList.clear();
+                    }
 
-            HistogramMetrics.Timer executeLatencyTimer = executeLatency.createTimer();
-            try {
-                List<List<PrepareRequest>> partitions = Lists.partition(prepareRequests, maxSyncoperationNum);
-                CountDownLatch countDownLatch = new CountDownLatch(partitions.size());
-                for (final List<PrepareRequest> partition : partitions) {
-                    executorService.submit(() -> {
+                    final List<PrepareRequest> finalPartition = partition;
+                    Future<?> submit = executorService.submit(() -> {
+                        HistogramMetrics.Timer executeLatencyTimer = executeLatency.createTimer();
                         try {
-                            if (CollectionUtils.isNotEmpty(partition)) {
-                                batchDAO.synchronous(partition);
+                            if (CollectionUtils.isNotEmpty(finalPartition)) {
+                                batchDAO.synchronous(finalPartition);
                             }
                         } catch (Throwable e) {
                             log.error(e.getMessage(), e);
                         } finally {
-                            countDownLatch.countDown();
+                            executeLatencyTimer.finish();
                         }
+
                     });
+                    results.add(submit);
                 }
-                countDownLatch.await();
-            } finally {
-                executeLatencyTimer.finish();
+
+                for (Future<?> result : results) {
+                    result.get();
+                }
+                return null;
+            });
+            countDownLatch.await();
+            prepareDone = true;
+            synchronized (prepareRequests) {
+                prepareRequests.notify();
             }
+            batchFuture.get();
+
         } catch (Throwable e) {
             errorCounter.inc();
             log.error(e.getMessage(), e);
         } finally {
             if (log.isDebugEnabled()) {
                 log.debug("Persistence data save finish");
             }
-
+            allTimer.finish();
             prepareRequests.clear();
+
             lastTime = System.currentTimeMillis();
+            prepareDone = false;
```

Comment:
Why need this?

Reply:
`prepareDone` means prepare stage is done. 
So we can't batch metrics to storage do not wait for metrics enoughly. And while loop for `batchDAO.synchronous` can break.
We can use local variables instead.